runtime/common/src/paras_registrar.rs

```diff
@@ -375,19 +375,19 @@ pub mod pallet {
 			<Self as Registrar>::apply_lock(para);
 			Ok(())
 		}
-
-		/// Schedule a parachain upgrade.
+	
+		/// Set the parachain's current code.
 		///
 		/// Can be called by Root, the parachain, or the parachain manager if the parachain is unlocked.
 		#[pallet::call_index(7)]
-		#[pallet::weight(<T as Config>::WeightInfo::schedule_code_upgrade(new_code.0.len() as u32))]
-		pub fn schedule_code_upgrade(
+		#[pallet::weight(<T as Config>::WeightInfo::set_current_code(new_code.0.len() as u32))]
+		pub fn set_current_code(
 			origin: OriginFor<T>,
 			para: ParaId,
 			new_code: ValidationCode,
 		) -> DispatchResult {
 			Self::ensure_root_para_or_owner(origin, para)?;
-			runtime_parachains::schedule_code_upgrade::<T>(para, new_code)?;
+			runtime_parachains::set_current_code::<T>(para, new_code)?;
```

Comment:
It is not that easy. As the origin can be `para_or_owner` which are not trusted. Thus, we still need to run the PVF pre checking by the validators. The only thing that should not be done is the setting of the `GoAhead` signal here: https://github.com/paritytech/polkadot/blob/ac65f3b3d65543b8602ec2435f71334a3c02e065/runtime/parachains/src/paras/mod.rs#L1366

https://github.com/paritytech/polkadot/blob/ac65f3b3d65543b8602ec2435f71334a3c02e065/runtime/parachains/src/paras/mod.rs#L1524-L1529 this is the piece of code that gets called after the pre-checking is done. One level above there is a pre-checking cause. This cause should include the info to not set the signal.

Before you do any of these changes, you should write an unit test to replicate this bug and then ensure that it works after your fixes.

Reply:
Makes sense. Thanks for the feedback, it is much appreciated. I will look into it and involve someone who can dedicate more time to working on this ;)